[shuyangk] iP

shuyangk opened the pull request:
# Charlie
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Charlie frees your mind of having to remember things you need to do. It's
- text-based
- intuitive
- ~~fast~~ * *SUPER FAST* * to use

All you need to do is download it from [here](https://github.com/shuyangk/ip/releases/tag/A-Jar).
1. double-click it.
2. add your tasks.
3. let it manage your tasks for you 😉
4. And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:

```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

shuyangk committed:
Level 0. Rename, Greet, Exit

shuyangk committed:
Level 1. Echo

shuyangk committed:
Level 2. Add, List

shuyangk committed:
Level 3. Mark as Done, added Task class

shuyangk committed:
added Deadline, ToDo, Event classes and functions

shuyangk committed:
implemented UI testing, fixed string displays

shuyangk committed:
added DukeException class, exception handling

shuyangk committed:
changed array of tasks to arraylist, added delete functionality.

shuyangk committed:
Added methods to write to text file, and read from text file.

shuyangk committed:
Added functionality for chatbot to recognise date and time in specific formats

shuyangk committed:
Revert "Added functionality for chatbot to recognise date and time in specific formats"

This reverts commit 8dba213f6347ae75270703d9f9c03c703d0d3ef5.

shuyangk committed:
Added functionality to read from and write to text file

shuyangk committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	data/duke.txt
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/ToDo.java

shuyangk committed:
Merge branch 'master' of https://github.com/shuyangk/ip

# Conflicts:
#	data/duke.txt
#	src/main/java/Event.java

shuyangk committed:
Added functionality for chatbot to recognise date formats

shuyangk committed:
Merge branch 'branch-Level-8'

shuyangk committed:
Added Parser, Storage, TaskList, Ui classes for OOP

shuyangk committed:
Merge branch 'A-MoreOOP'

shuyangk committed:
Packaged classes

shuyangk committed:
Merge branch 'A-Packages'

shuyangk committed:
Merge remote-tracking branch 'origin/add-gradle-support'

shuyangk committed:
Edited mainClass name

shuyangk committed:
Added JUnit tests for a few classes

shuyangk committed:
Merge branch 'A-JUnit'

shuyangk committed:
Created JAR file

shuyangk committed:
Merge branch 'A-Jar'

shuyangk committed:
Format code to follow coding standard

shuyangk committed:
Add JavaDocs to the code

shuyangk committed:
Merge branch 'A-CodingStandard'

shuyangk committed:
Merge branch 'A-JavaDoc'

# Conflicts:
#	src/main/java/duke/data/TaskList.java
#	src/main/java/duke/data/task/Task.java
#	src/main/java/duke/ui/Ui.java

shuyangk committed:
Minor fixes

shuyangk committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java

KumChaiYin commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,137 @@
+            while (s.hasNext()) {
+                String nextTask = s.nextLine();
+                String[] input = nextTask.split(" \\| ");
+                String category = input[0];
```
Maybe can use name to explain more?
```suggestion
                String taskCategory = input[0];
```

KumChaiYin commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,105 @@
+                    return;
+                }
+                case "list": {
+                    this.ui.printList(tasklist.getTaskList());
```
As other lines you use this.tasklist, maybe here should be same?
```suggestion
                    this.ui.printList(this.tasklist.getTaskList());
```

Cloud7050 replied:
Indeed, sometimes `this.` is used while other times it is not (for example, in the `Duke` constructor above). Perhaps something to consider.

KumChaiYin commented on src/main/java/duke/data/task/Deadline.java:
```diff
@@ -0,0 +1,35 @@
+        super(description);
+        this.deadline = deadline;
+        try {
+            LocalDate d1 = LocalDate.parse(deadline);
```
Instead of d1, maybe can use dueDate? (use name to explain)
```suggestion
            LocalDate dueDate = LocalDate.parse(deadline);
```

KumChaiYin commented on src/main/java/duke/data/task/Event.java:
```diff
@@ -0,0 +1,42 @@
+        this.start = start;
+        this.end = end;
+        try {
+            LocalDate d1 = LocalDate.parse(start);
+            LocalDate d2 = LocalDate.parse(end);
```
Same issue here.
```suggestion
            LocalDate startDate = LocalDate.parse(start);
            LocalDate endDate = LocalDate.parse(end);
```

Cloud7050 replied:
I would agree, maybe `d1` `d2` etc could be named differently?

Cloud7050 replied:
![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/44526554/75b1e3a0-ce35-4d49-81d0-24a9af4b87fb)

KumChaiYin commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    public void printAddedToListMessage(Task task, int size) {
+        System.out.println("____________________________________________________________\n" +
```
Perhaps use the dottedLine?
```suggestion
        System.out.println(dottedLine +
```

KumChaiYin commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,94 @@
+    }
+    private String dottedLine = "____________________________________________________________\n";
+
+    public void showLine() {
```
This method seems like didn't be used.

KumChaiYin reviewed (commented):
Looks good.
Overall, the code follows the coding standard, no issues with method and Boolean name. Can consider be more OOP by adding Command class to make the code clearer.

Cloud7050 commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    @Test
+    public void parser_getDeadlineDetail_success() throws DukeException {
+        assertEquals("this", new Parser().getDeadlineDescription("deadline this /by now"));
```
The code mainly has mentions of "description", but in the test method name, it is referred to as "detail". Should the test method name be updated?

Cloud7050 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,105 @@
+public class Duke {
+
+    private final Storage storage;
+    private TaskList tasklist;
```
Should the names `tasklist` in the various files be refactored to `taskList` to follow camel case?

Cloud7050 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    public void greet() {
+        System.out.println(dottedLine +
+                "Hello! I'm Charlie\n" +
+                "What can I do for you?\n" +
+                dottedLine);
```
Nice job following the 8-space indentation guideline for wrapped lines. However, there is another, intermediate rule that could apply:

![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/44526554/4699997a-600e-4cf8-8faf-4abad22821fc)

```suggestion
        System.out.println(dottedLine
                + "Hello! I'm Charlie\n"
                + "What can I do for you?\n"
                + dottedLine);
```

Cloud7050 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,94 @@
+
+
+public class Ui {
+    public Ui() {
+    }
```
Could this empty constructor be omitted?

Cloud7050 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,105 @@
+            String input = scanner.nextLine();
+            String keyword = parser.getInstruction(input);
+            try {
+                switch (keyword) {
+                case "bye": {
```
Great that you followed the indentation guideline for switch-case.

Cloud7050 commented on src/main/java/duke/data/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+
+public class TaskList {
```
Should this public class have a descriptive header comment?

Cloud7050 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.ArrayList;
+import duke.data.task.Task;
+
+
+public class Ui {
```
Similarly, should this public class have an associated Javadoc?

shuyangk committed:
Add check-style into the code, changed code according to it.

shuyangk committed:
Merge branch 'A-CheckStyle'

shuyangk committed:
Add GUI for chatbot

shuyangk committed:
Merge branch 'Level-10'

shuyangk committed:
Generate new JAR file using Gradle

shuyangk committed:
Add assertions into the code

Current code does not have assertions included.

Adding assertions will allow the code to handle unforeseen situations and scenarios
that should not occur at any point of the code by actively checking various states.

shuyangk committed:
Improve Code Quality

There are several methods that check if a task was marked or not. In some
cases the conversion is in-lined as a code block in another method.

Unifying all those duplicated code improves the code quality.
Improving code quality allows for better readability and maintainability
in the future.

shuyangk committed:
Merge pull request #1 from shuyangk/A-Assertions

Add assertions into the code

shuyangk committed:
Merge branch 'master' into A-CodeQuality

shuyangk committed:
Merge pull request #2 from shuyangk/A-CodeQuality

Improve Code Quality

shuyangk committed:
Add function to view schedule

There was no ommand for the user to find tasks
that occured on certain dates.

Adding such a function gives the user more flexibility in
searching from tasks apart from filtering by task name alone.

shuyangk committed:
Merge branch 'B-ViewSchedule'

shuyangk committed:
Fix issue where list is compressed when there are multiple tasks

The full list of tasks was compressed instead of displaying every
item in the list.

Displaying the full list allows users to view items that are at the
bottom as well, making the bot more user-friendly

shuyangk committed:
Replace avatar icons of user and bot

The previous avatar for the user and the bot was the default one
provided.

The new avatar matches the name of the chatbot

shuyangk committed:
Improve GUI

Background colour of user input box was previously white.

Having the same background colour as the rest of the window might
make it hard for users to spot.
Changing the background colour provides contrast and allows users to
spot it more easily.

shuyangk committed:
Merge branch 'A-BetterGui'

shuyangk committed:
Add personality of Charlie Brown to the bot

The bot responds in a regular manner.

The responses of the bot should represent the avatar representing it.

Changing the way the bot responds by including often used phrases by
Charlie Brown makes the bot have a personality similar to him.

shuyangk committed:
Merge branch 'A-Personality'

shuyangk committed:
Add more JUnit tests

There are new methods in the code that the previous JUnit tests
did not cover.

Adding these tests will check if the methods are working as intended
for future iterations.

shuyangk committed:
Merge branch 'A-MoreTesting'

shuyangk committed:
Improve JavaDocs

shuyangk committed:
Added User Guide

shuyangk committed:
Fix User Guide formatting

shuyangk committed:
Fixed User Guide formatting issue

shuyangk committed:
Fix bug where directory is not created if it does not exist

shuyangk committed:
Add release link to User Guide

shuyangk committed:
Update project README
